src/state/slices/tradeInputSlice/selectors.ts

```diff
@@ -224,3 +223,93 @@ export const selectHasUserEnteredAmount = createSelector(
+      : undefined
+    // Return the "best" quote even if it has errors, provided there is a quote to display data for
+    // this allows users to explore trades that aren't necessarily actionable. The UI will prevent
+    // executing these downstream.
+    const isSelectable = bestQuote?.quote !== undefined
+    const defaultQuoteMeta = isSelectable ? bestQuoteMeta : undefined
+    return tradeQuoteSlice.activeQuoteMeta ?? defaultQuoteMeta
+  },
+)
+
+const selectActiveSwapperApiResponse: Selector<ReduxState, ApiQuote | undefined> =
+  createDeepEqualOutputSelector(
+    selectTradeQuotes,
+    selectActiveQuoteMeta,
+    (tradeQuotes, activeQuoteMeta) => {
+      // If the active quote was reset, we do NOT want to return a stale quote as an "active" quote
+      if (activeQuoteMeta === undefined) return undefined
+
+      return tradeQuotes[activeQuoteMeta.swapperName]?.[activeQuoteMeta.identifier]
+    },
+  )
+const selectConfirmedQuote: Selector<ReduxState, TradeQuote | undefined> =
+  createDeepEqualOutputSelector(selectTradeQuoteSlice, tradeQuote => tradeQuote.confirmedQuote)
+const selectActiveQuote: Selector<ReduxState, TradeQuote | undefined> =
+  createDeepEqualOutputSelector(
+    selectActiveSwapperApiResponse,
+    selectConfirmedQuote,
+    (response, confirmedQuote) => {
+      // Return the confirmed quote for trading, if it exists.
+      // This prevents the quote changing during trade execution, but has implications on the UI
+      // displaying stale data. To prevent displaying stale data, we must ensure to clear the
+      // confirmedQuote when not executing a trade.
+      if (confirmedQuote) return confirmedQuote
+      return response?.quote
+    },
+  )
+
+const selectSecondHop: Selector<ReduxState, TradeQuote['steps'][number] | undefined> =
+  createDeepEqualOutputSelector(selectActiveQuote, quote => (quote ? quote.steps[1] : undefined))
+
+export const selectIsActiveQuoteMultiHop: Selector<ReduxState, boolean | undefined> =
+  createSelector(selectActiveQuote, quote => (quote ? quote?.steps.length > 1 : undefined))
+
+// if multi-hop, selects the account ID we're selling from for fee asset of the last hop and its account number
+// else, selects none because there's obviously no AccountId if there's no hop
+export const selectSecondHopSellAccountId = createSelector(
+  selectAccountIdByAccountNumberAndChainId,
+  selectIsActiveQuoteMultiHop,
+  selectSecondHop,
+  (accountIdsByAccountNumberAndChainId, isMultiHopTrade, secondHop) => {
+    // No second hop sellAccountId if there is no second hop
+    if (!isMultiHopTrade || !secondHop) return
+    const secondHopSellAssetAccountNumber = secondHop.accountNumber
+    const secondHopSellAssetChainId = secondHop.sellAsset.chainId
+
+    const chainIdAccountNumbers =
+      accountIdsByAccountNumberAndChainId[secondHopSellAssetAccountNumber]
+    if (!chainIdAccountNumbers) return
+    return chainIdAccountNumbers?.[secondHopSellAssetChainId]
```

Comment:
The guts of this PR, all the rest is re-declarations against circular deps shenanigans.
To understand why this was wrong, see:

https://github.com/shapeshift/web/blob/2c882b66aebec29bfd4b5960bf3cdd1fdc75dc37/src/state/slices/tradeInputSlice/selectors.ts#L102-L105

and https://github.com/shapeshift/web/blob/adfd1dfaf8b74a35cef4b092e750b57f5f2f3709/src/state/apis/swapper/helpers/validateTradeQuote.ts#L153-L158

Meaning for e.g a MATIC -> AVAX swap, we would call `selectPortfolioCryptoPrecisionBalanceByFilter` with AVAX as an AssetId (correct) but a MATIC AccountId (incorrect), failing balance checks since the interesection of those two obviously yields no account, therefore returning 0 balance and resulting in "Not enough AVA.AVAX to cover gas" error despite the AVAX balance being more than enough for gas.